### PR TITLE
Feature/sequence number in payload

### LINKED ISSUE #378: Append sequence in encryption payload
Instead of guessing the nonce append it to the payload.

### PR DESCRIPTION
**Pull request by ivcosla**
Closes #378

### CODE REVIEW

File: pkg/messaging/channel_test.go [hunk 1]
```diff
@@ -85,15 +85,15 @@ func TestChannelWrite(t *testing.T) {
...
+	buf := make([]byte, 29)
 	go out.Read(buf) // nolint
 	n, err := c.Write([]byte("foo"))
 	require.NoError(t, err)
 	assert.Equal(t, 3, n)
 
 	assert.Equal(t, FrameTypeSend, FrameType(buf[2]))
 	assert.Equal(t, byte(10), buf[3])
-	require.Equal(t, uint16(19), binary.BigEndian.Uint16(buf[4:]))
+	//require.Equal(t, uint16(19), binary.BigEndian.Uint16(buf[4:]))
```
Comment (ivcosla) [inline pkg/messaging/channel_test.go]: Not sure what this was testing, it was failing now.

File: internal/noise/noise.go [hunk 2]
```diff
@@ -31,8 +29,9 @@ type Noise struct {
 	enc     *noise.CipherState
 	dec     *noise.CipherState
 
-	encN uint32 // counter to inform encrypting CipherState to re-key
-	decN uint32 // counter to inform decrypting CipherState to re-key
+	seq uint32 // sequence number, used as nonce for both encrypting and decrypting
```
Comment (ivcosla) [inline internal/noise/noise.go]: Note that now there is no rekey (which was leading to chacha errors). Seq value is also not being set to zero before it overflows.

**Comment by ivcosla** [comment]
Tested using ssh integration environment. After running integration-startup I was able to run ssh without a single bug, however, ssh-integration-test failed to start ssh client correctly.

No data races were spotted during the test.

File: internal/noise/noise.go (hunk 2 above)
Comment (evanlinjin) [reply internal/noise/noise.go]: @ivcosla correct me if I'm wrong, shouldn't we have two `seq` numbers; one for encrypting `CipherState` and one for decrypting `CipherState`?

Comment (ivcosla) [reply internal/noise/noise.go]: I didn't saw the need for that. We are taking the seq number from the payload when decrypting, so how would it be used?. We can keep track of how many decryptions we have performed with it, but it won't be used for encryption right?. If you have some suggestion regarding this, go for it!.

**Review by ayuryshev — APPROVED** [review]
It's obviously improves life of ssh-sessions.
Actually I tried to break it and was unsuccessful in this attempt.

That's great!

File: internal/noise/noise.go (hunk 2 above)
Comment (evanlinjin) [reply internal/noise/noise.go]: Ahhh I see. However this implementation way lead to security issues. We need to test for non-duplicate nonces before decrypting.

Comment (evanlinjin) [reply internal/noise/noise.go]: As mentioned on Discord: We need to keep track of decryption nonces, and always ensure the next nonce > current nonce.

**Comment by ivcosla** [comment]
```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9bafe8]

goroutine 97 [running]:
github.com/flynn/noise.(*CipherState).Cipher(...)
        /home/ivan/Escritorio/skycoin/pkg/mod/github.com/flynn/noise@v0.0.0-20180327030543-2492fe189ae6/state.go:61
github.com/skycoin/skywire/internal/noise.(*Noise).EncryptUnsafe(0xc000140750, 0xc0002a8380, 0x208, 0x30a, 0xc0002a2120, 0xc0002978f0, 0x589e93)
        /home/ivan/skycoin/skywire/internal/noise/noise.go:129 +0x1b8
github.com/skycoin/skywire/pkg/messaging.(*msgChannel).Write(0xc000106630, 0xc0002a8380, 0x208, 0x30a, 0x0, 0x0, 0x99c314f9b87fe5f2)
        /home/ivan/skycoin/skywire/pkg/messaging/channel.go:131 +0x310
encoding/json.(*Encoder).Encode(0xc000297ef0, 0xbdfdc0, 0xc000097860, 0x7fc4883bf320, 0xc000106630)
        /usr/lib/go/src/encoding/json/stream.go:225 +0x241
github.com/skycoin/skywire/pkg/transport.settlementInitiatorHandshake.func1(0xc00019ad90, 0xd59ea0, 0xc000106630, 0xc0003b41e0, 0x0, 0xc00006fce0)
        /home/ivan/skycoin/skywire/pkg/transport/handshake.go:46 +0x8d2
github.com/skycoin/skywire/pkg/transport.settlementHandshake.Do.func1(0xc0003deb00, 0xc00019ad90, 0xd59ea0, 0xc000106630, 0xc000010890, 0xc000139760, 0xc0003dc720)
        /home/ivan/skycoin/skywire/pkg/transport/handshake.go:18 +0x56
created by github.com/skycoin/skywire/pkg/transport.settlementHandshake.Do
        /home/ivan/skycoin/skywire/pkg/transport/handshake.go:17 +0x108
```

I ran into this once when closing node B. Not sure what can be nil by then on line 129.

**Comment by ivcosla** [comment]
The referenced line in previous comment is actually this one: 
https://github.com/skycoin/skywire/pull/379/files#diff-9fba0b720ed6c60cb8dd622f7dc3edecR125

**Comment by jdknives** [comment]
This also closes #355 ?
